### PR TITLE
Upgrade inspector session

### PR DESCRIPTION
**Pull request by guysalt**

### CODE REVIEW

File: pymobiledevice3/lockdown.py [hunk 1]
```diff
@@ -103,7 +101,7 @@ class LockdownClient(object):
     def __init__(self, serial: str = None, hostname: str = None, client_name: str = DEFAULT_CLIENT_NAME,
                  autopair: bool = True, usbmux_connection_type: str = None, pair_timeout: int = None,
                  local_hostname: str = None, pair_record: Mapping = None,
-                 pairing_records_cache_folder: Path = HOMEFOLDER):
+                 pairing_records_cache_folder: Path = get_home_folder()):
```
Comment (matan1008) [inline pymobiledevice3/lockdown.py]: `get_home_folder` will be called on each import, you better put the default `None` and call it inside the function.
Also, remove the mkdir from line 546/544

File: pymobiledevice3/common.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from pathlib import Path
+
+_HOMEFOLDER = Path.home() / '.pymobiledevice3'
+
+
+def get_home_folder() -> Path:
+    _HOMEFOLDER.mkdir(exist_ok=True)
```
Comment (matan1008) [inline pymobiledevice3/common.py]: Add parents_ok

File: pymobiledevice3/services/web_protocol/inspector_session.py [hunk 3]
```diff
@@ -1,15 +1,35 @@
...
+
+    def __init__(self, dict_key):
+        self.dict_key = dict_key
+
+    def __repr__(self):
+        return str(self.dict_key)
+
+
 class InspectorSession:
+    __ERROR_SUPPORT_MESSAGE = 'Message is not supported'
```
Comment (matan1008) [inline pymobiledevice3/services/web_protocol/inspector_session.py]: Remove leading `_`

File: pymobiledevice3/cli/webinspector.py [hunk 4]
```diff
@@ -17,7 +17,6 @@
 
 logger = logging.getLogger(__name__)
 
-
```
Comment (matan1008) [inline pymobiledevice3/cli/webinspector.py]: Why?

File: pymobiledevice3/cli/webinspector.py [hunk 5]
```diff
@@ -175,7 +174,9 @@ async def inspector_js_loop(inspector: WebinspectorService, app: Application, pa
     while True:
         exp = input('> ')
         try:
-            print(await inspector_session.runtime_evaluate(exp))
+            result = await inspector_session.runtime_evaluate(exp)
+            if result:
```
Comment (matan1008) [inline pymobiledevice3/cli/webinspector.py]: Why?

File: pymobiledevice3/services/web_protocol/inspector_session.py [hunk 6]
```diff
@@ -97,3 +121,71 @@ async def receive_message(self, message_id: int) -> Mapping:
                 return response
 
             self._responses_cache[receive_message_id] = response
+
+    # -- PARSE OBJECTS
+    def _parse_target_dispatch_message(self, response: Mapping):
+        print(response)
```
Comment (matan1008) [inline pymobiledevice3/services/web_protocol/inspector_session.py]: Remove

File: pymobiledevice3/cli/webinspector.py [hunk 7]
```diff
@@ -6,17 +6,27 @@
...
 
 from pymobiledevice3.cli.cli_common import Command, wait_return
+from pymobiledevice3.common import get_home_folder
 from pymobiledevice3.exceptions import WirError, InspectorEvaluateError
 from pymobiledevice3.lockdown import LockdownClient
 from pymobiledevice3.services.web_protocol.cdp_server import app
 from pymobiledevice3.services.web_protocol.driver import WebDriver, Cookie, By
 from pymobiledevice3.services.webinspector import WebinspectorService, SAFARI, Application, Page
 
+WEBINSPECTOR_HISTORY_PATH = get_home_folder() / 'webinspector_history'
```
Comment (matan1008) [inline pymobiledevice3/cli/webinspector.py]: This will create the home folder on import

File: pymobiledevice3/cli/webinspector.py [hunk 8]
```diff
@@ -171,16 +181,27 @@ def automation_jsshell(lockdown: LockdownClient, url, timeout):
...
+                colorful_result = highlight(f'{result}', lexers.JavascriptLexer(),
+                                            formatters.TerminalTrueColorFormatter(style='stata-dark'))
+                print(colorful_result, end='')
+
+        except (KeyboardInterrupt, InspectorEvaluateError, NotImplementedError):  # KeyboardInterrupt Control-C
             pass
+        except EOFError:  # Control-D
+            return
+
+        await asyncio.sleep(0)
```
Comment (matan1008) [inline pymobiledevice3/cli/webinspector.py]: Why?

File: pymobiledevice3/__main__.py [hunk 9]
```diff
@@ -88,6 +88,9 @@ def cli():
         return
     except BrokenPipeError:
         traceback.print_exc()
+    except LaunchApplicationTimeoutError:
```
Comment (matan1008) [inline pymobiledevice3/__main__.py]: The main try except handles more low level exception, you should it put it under webinspector

File: pymobiledevice3/services/webinspector.py [hunk 10]
```diff
@@ -8,7 +8,8 @@
 
 import nest_asyncio
 
-from pymobiledevice3.exceptions import WebInspectorNotEnabledError, RemoteAutomationNotEnabled
+from pymobiledevice3.exceptions import WebInspectorNotEnabledError, RemoteAutomationNotEnabled, \
```
Comment (matan1008) [inline pymobiledevice3/services/webinspector.py]: `WebInspectorNotEnabledError` and `RemoteAutomationNotEnabled` does not exist in this commit

File: pymobiledevice3/__main__.py [hunk 11]
```diff
@@ -90,7 +90,8 @@ def cli():
         traceback.print_exc()
     except LaunchApplicationTimeoutError:
         logger.error('Unable to launch application (try to unlock device)')
-
+    except WebInspectorNotEnabledError:
```
Comment (matan1008) [inline pymobiledevice3/__main__.py]: The main try except handles more low level exception, you should it put it under webinspector

File: pymobiledevice3/__main__.py [hunk 12]
```diff
@@ -92,6 +93,8 @@ def cli():
         logger.error('Unable to launch application (try to unlock device)')
     except WebInspectorNotEnabledError:
         logger.error('Web inspector is not enable')
+    except RemoteAutomationNotEnabled:
```
Comment (matan1008) [inline pymobiledevice3/__main__.py]: The main try except handles more low level exception, you should it put it under webinspector

File: pymobiledevice3/cli/webinspector.py [hunk 13]
```diff
@@ -165,6 +165,7 @@ def automation_jsshell(lockdown: LockdownClient, url, timeout):
     inspector.connect(timeout)
     safari = inspector.open_app(SAFARI)
     session = inspector.automation_session(safari)
+
```
Comment (matan1008) [inline pymobiledevice3/cli/webinspector.py:173]: Why?

File: pymobiledevice3/lockdown.py (hunk 1 above)
Comment (guysalt) [reply pymobiledevice3/lockdown.py]: done

File: pymobiledevice3/common.py (hunk 2 above)
Comment (guysalt) [reply pymobiledevice3/common.py]: add `parents=True`

File: pymobiledevice3/services/web_protocol/inspector_session.py (hunk 3 above)
Comment (guysalt) [reply pymobiledevice3/services/web_protocol/inspector_session.py]: done

File: pymobiledevice3/cli/webinspector.py (hunk 4 above)
Comment (guysalt) [reply pymobiledevice3/cli/webinspector.py]: mistake, changed

File: pymobiledevice3/cli/webinspector.py (hunk 5 above)
Comment (guysalt) [reply pymobiledevice3/cli/webinspector.py]: changed to `if exp and not exp.isspace():`, to not evaluate if the the line is empty (like safari inspector behaviour)

File: pymobiledevice3/services/web_protocol/inspector_session.py (hunk 6 above)
Comment (guysalt) [reply pymobiledevice3/services/web_protocol/inspector_session.py]: im not seeing it right now, think i already removed it

File: pymobiledevice3/cli/webinspector.py (hunk 8 above)
Comment (guysalt) [reply pymobiledevice3/cli/webinspector.py]: removed

File: pymobiledevice3/cli/webinspector.py (hunk 13 above)
Comment (guysalt) [reply pymobiledevice3/cli/webinspector.py:173]: mistake, removed

File: pymobiledevice3/__main__.py (hunk 12 above)
Comment (guysalt) [reply pymobiledevice3/__main__.py]: changed to webinspector decorator

File: pymobiledevice3/__main__.py (hunk 11 above)
Comment (guysalt) [reply pymobiledevice3/__main__.py]: changed to webinspector decorator

File: pymobiledevice3/__main__.py (hunk 9 above)
Comment (guysalt) [reply pymobiledevice3/__main__.py]: changed to webinspector decorator

File: pymobiledevice3/cli/webinspector.py (hunk 7 above)
Comment (guysalt) [reply pymobiledevice3/cli/webinspector.py]: move variable inside the function, hopefully i will change it in the future

File: pymobiledevice3/services/web_protocol/inspector_session.py [hunk 14]
```diff
@@ -96,4 +89,20 @@ async def receive_message(self, message_id: int) -> Mapping:
             if receive_message_id == message_id:
                 return response
 
-            self._responses_cache[receive_message_id] = response
+    @staticmethod
+    def _parse_runtime_evaluate(response: Mapping):
+        message = json.loads(response['params']['message'])
+        if message.get('error', None):
```
Comment (matan1008) [inline pymobiledevice3/services/web_protocol/inspector_session.py]: `'error ' not in message` or `message.get('error', None) is not None`

File: pymobiledevice3/cli/webinspector.py [hunk 15]
```diff
@@ -178,6 +178,8 @@ async def inspector_js_loop(inspector: WebinspectorService, app: Application, pa
             print_json(await inspector_session.runtime_evaluate(exp))
         except InspectorEvaluateError:
             pass
+        except NotImplementedError:
```
Comment (matan1008) [inline pymobiledevice3/cli/webinspector.py]: Why?

File: pymobiledevice3/services/web_protocol/inspector_session.py [hunk 16]
```diff
@@ -83,12 +94,12 @@ async def _receive_loop(self):
...
-            if response_method == 'Target.dispatchMessageFromTarget':
-                self._target_dispatch_message_from_target(response)
+            if response_method in self.response_methods:
+                self.response_methods[response_method](response)
             else:
-                error_message = 'Message is not supported'
+                error_message = 'Unknown response method'
                 logger.error(error_message, response)
-                raise NotImplementedError(error_message)
+                raise RuntimeError(error_message)
```
Comment (matan1008) [inline pymobiledevice3/services/web_protocol/inspector_session.py]: `RuntimeError` is very implicit, better using `NotImplementedError` or a custom exception

File: pymobiledevice3/services/webinspector.py [hunk 17]
```diff
@@ -191,7 +192,10 @@ def get_open_pages(self):
     def open_app(self, bundle: str, timeout: Union[float, int] = 3) -> Application:
         self.await_(self._request_application_launch(bundle))
         self.get_open_pages()
-        return self.await_(asyncio.wait_for(self._wait_for_application(bundle), timeout=timeout))
+        try:
+            return self.await_(asyncio.wait_for(self._wait_for_application(bundle), timeout=timeout))
+        except TimeoutError:
+            raise LaunchApplicationTimeoutError()
```
Comment (matan1008) [inline pymobiledevice3/services/webinspector.py]: Too specific, better name it `LaunchingApplicationError `

File: pymobiledevice3/services/web_protocol/inspector_session.py [hunk 18]
```diff
@@ -76,6 +76,9 @@ async def runtime_evaluate(self, exp: str):
 
         return self._parse_runtime_evaluate(response)
 
+    async def navigate_to_url(self, url: str):
+        return await self.runtime_evaluate(exp=f"window.location = '{url}'")
```
Comment (matan1008) [inline pymobiledevice3/services/web_protocol/inspector_session.py]: `f'window.location = "{url}"'`

File: pymobiledevice3/cli/webinspector.py [hunk 19]
```diff
@@ -229,27 +231,27 @@ async def inspector_js_loop(inspector: WebinspectorService, app: Application, pa
...
+@click.argument('url', required=False, default='')
 @catch_errors
-def inspector_jsshell(lockdown: LockdownClient, timeout):
+def js_shell(lockdown: LockdownClient, timeout, automation, url):
     """
     Opt in:
 
         Settings -> Safari -> Advanced -> Web Inspector
+
+        for automation also enable:
```
Comment (matan1008) [inline pymobiledevice3/cli/webinspector.py]: Unindent

File: pymobiledevice3/services/web_protocol/inspector_session.py (hunk 14 above)
Comment (guysalt) [reply pymobiledevice3/services/web_protocol/inspector_session.py]: changed to `'error' in message`

File: pymobiledevice3/cli/webinspector.py (hunk 15 above)
Comment (guysalt) [reply pymobiledevice3/cli/webinspector.py]: removed.

File: pymobiledevice3/services/web_protocol/inspector_session.py (hunk 16 above)
Comment (guysalt) [reply pymobiledevice3/services/web_protocol/inspector_session.py]: removed.

File: pymobiledevice3/services/webinspector.py (hunk 17 above)
Comment (guysalt) [reply pymobiledevice3/services/webinspector.py]: changed

File: pymobiledevice3/services/web_protocol/inspector_session.py (hunk 18 above)
Comment (guysalt) [reply pymobiledevice3/services/web_protocol/inspector_session.py]: changed

File: pymobiledevice3/cli/webinspector.py (hunk 19 above)
Comment (guysalt) [reply pymobiledevice3/cli/webinspector.py]: changed

File: pymobiledevice3/services/web_protocol/inspector_session.py [hunk 20]
```diff
@@ -18,7 +18,16 @@ def __init__(self, protocol: SessionProtocol, target_id: str):
...
+        self._dispatch_message_responses = {}
+
+        self.response_methods = {
+            'Target.targetCreated': self._target_created,
+            'Target.targetDestroyed': self._target_destroyed,
+            'Target.dispatchMessageFromTarget': self._target_dispatch_message_from_target,
+            'Target.didCommitProvisionalTarget': self._target_did_commit_provisional_target,
+        }
+
+        asyncio.create_task(self._receive_loop())
```
Comment (matan1008) [inline pymobiledevice3/services/web_protocol/inspector_session.py]: save reference

Comment (guysalt) [reply pymobiledevice3/services/web_protocol/inspector_session.py]: done

**Pull request merged**